test/Microsoft.ML.Functional.Tests/Explainability.cs

```diff
@@ -0,0 +1,284 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Microsoft.Data.DataView;
+using Microsoft.ML.Data;
+using Microsoft.ML.Functional.Tests.Datasets;
+using Microsoft.ML.RunTests;
+using Microsoft.ML.TestFramework;
+using Microsoft.ML.Trainers;
+using Microsoft.ML.Trainers.FastTree;
+using Xunit;
+using Xunit.Abstractions;
+
+namespace Microsoft.ML.Functional.Tests
+{
+    /// <summary>
+    /// Test explainability features.
+    /// </summary>
+    public class Explainability : BaseTestClass
+    {
+        public Explainability(ITestOutputHelper output) : base(output)
+        {
+        }
+
+        /// <summary>
+        /// GlobalFeatureImportance: PFI can be used to compute global feature importance.
+        /// </summary>
+        [Fact]
+        public void GlobalFeatureImportanceWithPermutationFeatureImportance()
+        {
+            var mlContext = new MLContext(seed: 1, conc: 1);
+
+            // Get the dataset
+            var data = mlContext.Data.ReadFromTextFile<HousingRegression>(GetDataPath(TestDatasets.housing.trainFilename), hasHeader: true);
+
+            // Create a pipeline to train on the housing data.
+            var pipeline = mlContext.Transforms.Concatenate("Features", HousingRegression.Features)
+                .Append(mlContext.Regression.Trainers.StochasticDualCoordinateAscent());
+
+            // Fit the pipeline and transform the data.
+            var model = pipeline.Fit(data);
+            var transformedData = model.Transform(data);
+
+            // Compute the permutation feature importance to look at global feature importance.
+            var permutationMetrics = mlContext.Regression.PermutationFeatureImportance(model.LastTransformer, transformedData);
```

Comment:
>permutationMetrics [](start = 16, length = 18)

any way to know if the PFI are actually correct - meaining that important features were marked as such?  Perhaps get a baseline or use generated dataset

Reply:
The tests for PFI correctness are in [`Microsoft.ML.Tests`](https://github.com/dotnet/machinelearning/blob/master/test/Microsoft.ML.Tests/PermutationFeatureImportanceTests.cs). They validate that the importances produced by PFI are correct.

The purpose of `Microsoft.ML.Functional.Tests` is to guarantee that end-to-end scenarios work, and that the results are returned in a way that make sense: That metrics objects are returned, that the individual metrics are in the allowable range for the metric, etc. In other words, these are not meant to be baseline or correctness tests.

---
In reply to: [258589829](https://github.com/dotnet/machinelearning/pull/2584#discussion_r258589829) [](ancestors = 258589829)

Reply:
This test would fail to detect if the method will return a totally invalid positive value via an API bug.

---
In reply to: [259082766](https://github.com/dotnet/machinelearning/pull/2584#discussion_r259082766) [](ancestors = 259082766,258589829)

Reply:
Offline conversation: We propose merging Functional tests and Baseline tests when we solve issue #2171, to move baseline tests off of the subcomponent infrastructure.

---
In reply to: [259084225](https://github.com/dotnet/machinelearning/pull/2584#discussion_r259084225) [](ancestors = 259084225,259082766,258589829)